### PR TITLE
feat(ir): stmt dependency analysis with InOut-use discipline check

### LINKED ISSUE #1027: [Feature] Dependency analysis with in-pass InOut-use check for LiftUnrolledLoads
### Summary

Implement the dependency analysis used by the `LiftUnrolledLoads` pass from RFC #1025 (partial loop unrolling for ping-pong buffering). The analysis answers "can this `tile.load` be hoisted past these intervening statements?" via standard SSA def-use, and enforces the **InOut-use discipline** defined in RFC #1026 as a precondition check at entry. If the input violates the discipline, the analysis rejects it with a precise diagnostic; otherwise, it returns a conservative-safe answer based on SSA def-use alone.

This is **Phase 1** of the rollout described in RFC #1026. The discipline is checked locally inside this analysis, not globally in the verifier. Phase 2 (tracked separately, see RFC #1026) will promote the check to a structural verifier at `VerificationLevel::Basic` and simplify this analysis by removing the embedded check.

### Motivation / Use Case

**Direct blocker for RFC #1025** (`PartialUnrollTileLoops` + `LiftUnrolledLoads`). The lift pass must decide, for each `tile.load` inside a marked-unrolled `SeqStmts`, whether it can be hoisted to the top of the region without violating dataflow. Two conflict modes exist in principle:

- **Mode A — SSA dataflow conflict:** the load reads a var that is defined by a skipped-over op.
- **Mode B — Physical memory aliasing:** skipped-over stores or Out/InOut calls mutate memory the load reads.

Under the InOut-use discipline (RFC #1026), Mode B is *subsumed* by Mode A: any memory mutation visible to a later load is mirrored by an SSA version change, so a load that would be affected by a mutation must read the post-mutation SSA var — detectable as a standard def-use dependency. This collapses the analysis from ~200 LOC of MemRef-based alias reasoning to ~50 LOC of SSA def-use + an entry-time discipline check.

Without this analysis, `LiftUnrolledLoads` cannot land, and `PartialUnrollTileLoops` produces IR that `MemoryReuse` (#22) coalesces back into a single buffer — defeating the ping-pong intent. So this issue is gating #1025.

### Details

**Scope:**
- New dependency-analysis entry point used by `LiftUnrolledLoads`.
- Embedded precondition check for the RFC #1026 discipline, scoped to the statements the analysis is invoked over.
- Lives initially inside `src/ir/transforms/lift_unrolled_loads_pass.cpp`. If a second caller emerges before Phase 2 lands, promote to `include/pypto/ir/transforms/utils/`.

**API sketch:**

```cpp
// src/ir/transforms/lift_unrolled_loads_pass.cpp

struct HoistabilityResult {
  bool can_hoist;
  // If false, optional reason for debug logging.
  std::optional<std::string> reason;
};

// Run on a marked SeqStmts. `load` is a candidate hoistable stmt; `skip_ops`
// are the stmts it would skip over if hoisted to the top.
//
// Preconditions: the caller has already validated (via a separate walk over
// the full unroll region) that the InOut-use discipline holds in this region.
// If the walk detected a violation, the caller reports the error and refuses
// to run the lift.
HoistabilityResult CanHoist(const StmtPtr& load,
                            const std::vector<StmtPtr>& skip_ops);

// Separate: validate the discipline on an entire unroll region. Reports
// violations via DiagnosticEmitter and returns true iff no violations.
bool ValidateInOutUseDiscipline(const StmtPtr& unroll_region);
```

**Implementation notes:**
- `CanHoist` iterates `skip_ops`; for each, collect defs via existing `VarDefUseCollector`; check membership against `CollectReads(load)`. No MemRef reasoning; the discipline guarantees this is sound.
- `ValidateInOutUseDiscipline` walks the region in CFG order, maintaining a "dead-for-read" set per var. On each call, add its InOut/Out-passed vars to the set (with scope tracking). On each read, check against the set. Report with precise source location on any intersection.
- Both helpers use PyPTO exceptions (`pypto::ValueError`) for hard rejects; debug logs for soft "cannot hoist this one" outcomes.

**Diagnostic example (on discipline violation):**

```
error: partial-unroll loop body violates InOut-use discipline
  --> user_kernel.py:42
   |
42 |     T_new = f(x, inout=T)
   |                 ^^^^^^^ `T` is InOut here
43 |     y = pl.tile.load(T, offset=i+32)
   |                       ^ `T` read after InOut call; use the return `T_new` instead
   |
help: rewrite as `y = pl.tile.load(T_new, offset=i+32)` inside the unrolled loop body
note: this rule is enforced because partial-unroll + ping-pong requires precise
      dataflow; see RFC #1026 for details
```

**Tests (required):**

- Unit tests covering:
  - Simple hoist (no InOut, no deps): succeeds.
  - Hoist past a call whose return is read by the load: rejected (Mode A).
  - Discipline violation (read of InOut-passed var): analysis rejects input with precise diagnostic.
  - Rule-conforming InOut usage (post-call reads go through returns): succeeds.
  - Nested-scope discipline violation (read in an `if` branch following the InOut call): rejected.
- Integration test: `pl.range(64, unroll=4)` on a discipline-conforming kernel produces the expected lifted-load IR and downstream ping-pong buffer allocation.
- Negative integration test: `pl.range(64, unroll=4)` on a violating kernel errors out with the precise diagnostic.

**Out of scope for this issue:**
- Promoting the discipline check to a structural verifier (Phase 2, tracked as the second implementation issue for RFC #1026).
- MemRef-based alias analysis as a fallback for discipline-violating input. If/when we have a reason to relax the discipline, this becomes a separate design discussion — not deferred work for this issue.
- Auditing existing kernels for discipline violations outside the partial-unroll path. That's a Phase 2 concern.

### Related

- Blocks: RFC #1025 (partial loop unrolling — this analysis is a direct dependency).
- Part of: RFC #1026 (InOut-use discipline, Phase 1).
- Follow-up: Phase 2 (structural verifier + simplify this analysis) — will be filed as a separate feature issue referencing RFC #1026.

### PR DESCRIPTION
**Pull request by Hzfengsy**
## Summary

Phase 1 of RFC #1026: adds a reusable statement dependency analysis utility plus the InOut-use discipline check that guarantees its soundness. Lands the building blocks that RFC #1025's `LiftUnrolledLoads` pass — and any future dataflow-aware pass — will consume.

- `BuildStmtDependencyGraph(region, program)` — SSA def→use DAG over a region's top-level statements. Compound children (If/For/...) aggregate their subtree's uses and defs so edges land at block granularity.
- `CheckInOutUseDiscipline(region, program)` — raises `VerificationError` when a variable that was passed as `InOut` or `Out` to a user-function call is read afterwards in CFG order. `Out` is treated identically to `InOut`: both mutate the caller's buffer, so post-call values must flow through the call's return slots.
- The graph builder runs the discipline check first (when a program is supplied), so callers cannot accidentally receive an unsound graph — a violating region halts compilation instead of silently returning wrong answers.
- If-branch snapshots prevent a post-call mark in one branch from bleeding into its sibling.
- Built-in ops (`tile.*`, `tensor.*`, `system.*`) do not contribute to the dead set — their memory mutations are Mode B in the RFC, out of scope for this dataflow analysis.
- Exposed via `passes.stmt_dependency_analysis` Python submodule for downstream consumers and tests.

## Test plan

- [x] 15 new unit tests covering graph construction (linear chains, independent stmts, compound children aggregating subtree use/def, non-SeqStmts regions, vars defined outside the region) and discipline enforcement (clean programs pass, post-call reads raise with the expected var name, `Out` parity with `InOut`, reads inside nested `if`/`for` raise, sibling-branch reads don't bleed, self-reads in the same call's args are allowed, built-in ops don't kill vars, `BuildStmtDependencyGraph` raises on violation)
- [x] Full `tests/ut/ir/` regression (`python -m pytest tests/ut/ir/ -n auto --maxprocesses 8` — 2520 passed, 13 skipped)
- [x] clang-tidy clean (`python tests/lint/clang_tidy.py --diff-base HEAD~2`)
- [x] Pre-commit hooks pass (clang-format, cpplint, ruff, pyright)

## Related Issues

Fixes #1027
Related: #1025, #1026

### CODE REVIEW

File: src/ir/transforms/utils/stmt_dependency_analysis.cpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+      VisitStmt(*op->else_body_);
+    }
+
+    // Merge: a var is dead after the if iff it was dead in either branch.
+    // Iteration order doesn't affect the result (insert into unordered_set is
+    // commutative and idempotent), so the range-insert form is deterministic.
+    dead_.insert(dead_after_then.begin(), dead_after_then.end());
+  }
+
+ private:
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/utils/stmt_dependency_analysis.cpp:217]: Fixed in `a7839670` — loop back-edges now modelled by pre-populating `dead_` with every var the body kills before walking the body. A read anywhere in the body of a var killed on some iteration is now flagged. Regression test: `test_loop_backedge_is_flagged`.

File: include/pypto/ir/transforms/utils/stmt_dependency_analysis.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+ *   - Otherwise the region itself is the single node.
+ *
+ * Edges are predecessor relationships at block granularity: a compound child
+ * (IfStmt/ForStmt/...) aggregates all the variable uses and defs from its
+ * subtree. `predecessors[X]` is the set of nodes whose defs X directly reads.
+ *
+ * The graph is sound under the InOut-use discipline (RFC #1026): physical
+ * memory mutation is mirrored by SSA version changes, so SSA def-use captures
+ * all real dependencies. Callers needing soundness must first run
+ * CheckInOutUseDiscipline and refuse to proceed on any violation.
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply include/pypto/ir/transforms/utils/stmt_dependency_analysis.h]: Fixed in `a7839670` — updated the `StmtDependencyGraph` struct docstring to note that `BuildStmtDependencyGraph` enforces the discipline precondition when a program is supplied.

File: src/ir/transforms/utils/stmt_dependency_analysis.cpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+      std::string origin_str =
+          origin_it != dead_origin_.end() ? origin_it->second.to_string() : std::string("<unknown>");
+      std::string msg = "variable '" + op->name_hint_ + "' was passed as InOut/Out at " + origin_str +
+                        "; read the post-call return value instead of the pre-call variable";
+      diagnostics_.emplace_back(DiagnosticSeverity::Error, "InOutUseDiscipline", 0, std::move(msg),
+                                op->span_);
+    }
+    // Delegate to base so IterArg::initValue_ is still visited.
+    IRVisitor::VisitVarLike_(op);
+  }
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/utils/stmt_dependency_analysis.cpp:129]: Fixed in `a7839670` — overrode stmt visitors (AssignStmt, IfStmt, ForStmt, WhileStmt) to descend only into read contexts (RHS/condition/body/iter-arg init values), skipping LHS, `loop_var_`, `iter_args_`, and `return_vars_`. Regression test: `test_for_loop_var_is_not_flagged`.

File: src/ir/transforms/utils/stmt_dependency_analysis.cpp [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    // commutative and idempotent), so the range-insert form is deterministic.
+    dead_.insert(dead_after_then.begin(), dead_after_then.end());
+  }
+
+ private:
+  ProgramPtr program_;
+  std::unordered_set<const Var*> dead_;
+  std::unordered_map<const Var*, Span> dead_origin_;
+  std::vector<Diagnostic> diagnostics_;
+};
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/utils/stmt_dependency_analysis.cpp:239]: Fixed in `a7839670` — added back-edge modelling to `ForStmt` and `WhileStmt`: before walking the body, we pre-populate `dead_` with the full set of vars the body would mark InOut/Out-dead. A read anywhere in the body of such a var is then flagged (captures the back-edge case where iteration N's call kills a var read in iteration N+1). Regression test: `test_loop_backedge_is_flagged`.

File: src/ir/transforms/utils/stmt_dependency_analysis.cpp [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    for (const auto& arg : op->args_) {
+      VisitExpr(arg);
+    }
+
+    // Resolve callee. Built-in ops (tile.*, tensor.*, system.*) won't resolve;
+    // they do not contribute to the dead set. Their memory mutations are
+    // handled as Mode B in RFC #1026, which is out of scope here.
+    if (!program_) return;
+    FunctionPtr callee = program_->GetFunction(op->op_->name_);
+    if (!callee) return;
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/utils/stmt_dependency_analysis.cpp:143]: Not changing — `Program::GetFunction` already returns null for names that aren't registered, and the subsequent `if (!callee) return;` skips the call. No inconsistency can slip through.

File: src/ir/transforms/utils/stmt_dependency_analysis.cpp [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    dead_ = std::move(snapshot);
+    if (op->else_body_.has_value()) {
+      VisitStmt(*op->else_body_);
+    }
+
+    // Merge: a var is dead after the if iff it was dead in either branch.
+    // Iteration order doesn't affect the result (insert into unordered_set is
+    // commutative and idempotent), so the range-insert form is deterministic.
+    dead_.insert(dead_after_then.begin(), dead_after_then.end());
+  }
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/transforms/utils/stmt_dependency_analysis.cpp:183]: Not changing — the current implementation is a single O(N) forward walk (one `VarDefUseCollector` per stmt), well within `.claude/rules/pass-complexity.md`. A backward suffix pre-compute would double the traversal cost without changing asymptotics. The one exception — loop body pre-scan for back-edge modelling — is handled in a separate commit.

**Review by lyfne123 — APPROVED** [review]
